server/src/main/java/org/elasticsearch/action/ActionRunnable.java

```diff
@@ -79,6 +80,36 @@ public String toString() {
         };
     }
 
+    /**
+     * Like {#wrap} except with a {@link Releasable} which is released after executing the consumer, or if the action is rejected.
+     */
+    public static <T> ActionRunnable<T> wrapReleasing(
```

Comment:
This assumes we would never call `onFailure` on a releasable other than through `onRejection` unless `doRun` was invoked before. I can't think of a counter example here but do we want to depend on this?
Any reason not to just put the release in the `onAfter` instead to get standard behaviour here?

Reply:
Yeah we want the release to happen right after calling the consumer even if it doesn't immediately complete the listener. Conceptually the consumer is `transportService#sendRequest` which acquires a ref and releases it once the request is sent - we don't want to otherwise delay the release of the request until the response comes back.

But yeah that's a good point about calling `onFailure` directly. Doesn't happen when submitting this to an executor, but I suppose this could be used in other situations too. I pushed 227d8f1deac6a4478294807088c3c099f77272b8.

Reply:
NB calling `onFailure` directly doesn't call `onAfter`. That might be surprising in other ways too.